app/code/community/Bolt/Boltpay/Model/Payment.php

```diff
@@ -547,6 +557,148 @@ public function handleTransactionUpdate(Mage_Payment_Model_Info $payment, $newTr
+                                $creditmemo = $service->prepareInvoiceCreditmemo($invoice, $data);
+                                $creditmemo->setRefundRequested(true);
+                                $creditmemo->setOfflineRequested(false);
+                                $creditmemo->setPaymentRefundDisallowed(false);
+                                $creditmemo->register()->save();
+                            }
+                        }  
+                    }
+                    $order->save();
+                    
+                    $baseTotalRefunded = $order->getBaseTotalRefunded();
+                    if($baseTotalRefunded){
+                        $baseAvailableRefund = Mage::app()->getStore()->roundPrice(
+                                $baseTotalPaid - $baseTotalRefunded
+                            );
+                    }
+                    else{
+                        $baseAvailableRefund = Mage::app()->getStore()->roundPrice(
+                                $baseTotalPaid
+                            );
+                    }
+
+                    if($baseAvailableRefund < 0.01){
+                        //for partial refund, after all the paid amount is refuned
+                        //we need to restore the items in cart separately
+                        if($isPartialRefund){
+                            foreach($invoiceIds as $k=>$invoiceId){
+                                $invoice = Mage::getModel('sales/order_invoice')
+                                            ->load($invoiceId)
+                                            ->setOrder($order);
+                                $qtys = array();
+                                foreach($order->getAllItems() as $item) {
+                                    $qtys[$item->getId()] = $item->getData('qty_ordered');
+                                }   
+                        
+                                $data = array(
+                                    'qtys' => $qtys
+                                );
+                                $creditmemo = $service->prepareInvoiceCreditmemo($invoice, $data);
+                                $creditmemo->setSubtotal(0);
+                                $creditmemo->setShippingAmount(0);
+                                $creditmemo->setBaseGrandTotal(0);
+                                $creditmemo->setGrandTotal(0);
+                                $creditmemo->setRefundRequested(false);
+                                $creditmemo->setOfflineRequested(false);
+                                $creditmemo->setPaymentRefundDisallowed(true);
+                                $creditmemo->register()->save();
+                            }
+                        }
+                        $payment->setIsTransactionClosed(true);
+                        $payment->setShouldCloseParentTransaction(true);          
+                        $order->save();
+                    }
+                }
```

Comment:
In looking at this entire section, my intuition is telling me that it is much to complex for a fairly straight-forward task.  If we are to break down the flow to be:

1.) get the order
2.) get the order's invoice which we know for Bolt should only be one
3.) create the credit memo to issue a refund

Then, perhaps the code for full and partial refunds could be consolidated and simplified into something like this:

```
} elseif ($prevTransactionStatus == 'preparing_to_refund') {
    $invoices = $order->getInvoiceCollection()->getItems();
    $invoice = reset($invoices);  // or $invoice = $invoices[0];
    $service = Mage::getModel('sales/service_order', $order);
    $creditmemo = $service->prepareInvoiceCreditmemo($invoice);
    $creditmemo->setAdjustment($transactionAmount);  // currently labelled $captureAmount
    $creditmemo->refund();
} elseif ($newTransactionStatus == self::TRANSACTION_COMPLETED) {
  ...
```
This way, 
 - you greatly simplify your code, making it easier to understand and update
 - you let Magento do all the crazy calculations in its own uniform way, 
 - you trigger all events automatically that are supposed to be triggered

Reply:
well, I do a lot of investigation here. 

1. for full refund I do think I can simplify it a bit.
```
$qtys = array();
                                foreach($order->getAllItems() as $item) {
                                    $qtys[$item->getId()] = $item->getData('qty_ordered');
                                }   
                        
                                $data = array(
                                    'qtys' => $qtys
                                );
```
that remove the snippet above. cause by default if we do full refund, the magento would take all the items

2. if we do partial refund, since it is hard to tell which item should be refund in this transaction, we have to set the qty to 0, if not, magento would refund full.